Parse escaped chars

schorschii opened the pull request:
Hi! First of all, thanks for this library!

I noticed that it does not parse escaped chars like `\n` or `\u00fc` (ü), see [here](https://stackoverflow.com/questions/19176024/how-to-escape-special-characters-in-building-a-json-string#answer-27516892).

This change adds support for those escaped values.

schorschii committed:
parse escaped chars

schorschii committed:
improve code

schorschii committed:
Parse escaped chars (#1)

schorschii committed:
add data parse and int convenience functions (#2)

mikezs commented on Source/JSON.swift:
```diff
@@ -596,7 +596,29 @@ open class JSON {
+            .replacingOccurrences(of: "\\t", with: "\t", range: nil)
+            .replacingOccurrences(of: "\\r", with: "\r", range: nil)
+            .replacingOccurrences(of: "\\\"", with: "\"", range: nil)
+        if let regex = try? NSRegularExpression(pattern: "\\\\u[0-9A-Fa-f]{4}") {
```
What's this block doing, it's not very descriptive and doesn't look like it relates to escaped characters, it looks like some sort of hex number parsing.

Edit: I see this is from the link you posted in the original PR, but the variable names don't make it obvious at all

mikezs commented on Source/JSON.swift:
```diff
@@ -596,7 +596,29 @@ open class JSON {
         }
         
-        let parsedString = String(jsonString[startingIndex ..< index])
+        var parsedString = String(jsonString[startingIndex ..< index])
```
This could stay as a constant and be given a different name and built upon, e.g. 
`let initialParsedString = String()`
`let parsedString = initialParsedString.someOperation()`

schorschii replied:
I'm not sure if this is useful. I mean how would you implement this for the `hexEncodedUnicodeCharRegex` `parsedString`? Feel free to push your solution on this branch.

mikezs commented:
I've added some suggestions, and please add some tests for each of the scenarios and edge cases please

schorschii committed:
add test

schorschii commented:
I added a test for `\n` and `\u00fc` and also fixed the `testEscapedQuoteString()` test:

```

let input = """
["\\\""]
"""
```
results in `["\""]`, which is just a `"` when decoding the JSON string. The previous version expected `\"` which is wrong.


> I've added some suggestions

What do you mean? I can't see any comments or changes on the code.

schorschii committed:
unify testEscapedQuoteString and testEscapedCharsString tests and add \t, \r chars to test

mikezs commented:
> What do you mean? I can't see any comments or changes on the code.

They're comments on the code made on this pull request, see here: https://github.com/mikezs/Tisander/pull/2

schorschii commented:
I'm sorry but I can't see any comments.

mikezs reviewed (changes requested):
Sorry, apparently I left this in draft

schorschii committed:
change variable name

schorschii committed:
parse escaped /

schorschii committed:
Merge branch 'master' into parse-escaped-chars

schorschii committed:
add support for escaped backslash as last char in a string

schorschii committed:
fix edge case where second last char is a backslash

schorschii committed:
revert changes which do not belong to this feature/PR
